### PR TITLE
FIRECERT-1789:[FCS] - Use window values for app transitions and language

### PR DESCRIPTION
**Pull request by SnehalP251**

### CODE REVIEW

File: cypress/support/cypress-commands/lifecycle.js [hunk 1]
```diff
@@ -477,3 +478,67 @@ Cypress.Commands.add('lifecycleSchemaChecks', (response, state) => {
...
+ * @module commands
+ * @function validateVisibilityState
+ * @description To validate app visibility for different lifecycle states
+ * @param {String} state - To check visibilityState of state
+ * @example
+ * validateVisibilityState('foreground');
+ */
+function validateVisibilityState(state) {
+  // Fetching the visibilityState for the states from env.
+  const visibilityState = Cypress.env('visibilityState');
```
Comment (kschrief) [inline cypress/support/cypress-commands/lifecycle.js:492]: "visibilityState" is used in multiple places and should be in a constant.

**Review by kschrief — CHANGES_REQUESTED** [review]

File: cypress/support/cypress-commands/lifecycle.js [hunk 2]
```diff
@@ -554,3 +555,70 @@ Cypress.Commands.add('lifecycleSchemaChecks', (response, state) => {
...
+ * @function validateVisibilityState
+ * @description To validate app visibility for different lifecycle states
+ * @param {String} state - To check visibilityState of state
+ * @example
+ * validateVisibilityState('foreground');
+ */
+function validateVisibilityState(state) {
+  // Fetching the visibilityState for the states from env.
+  const visibilityState = Cypress.env(CONSTANTS.VISIBILITYSTATE);
+  if (visibilityState.hasOwnProperty(state)) {
```
Comment (kschrief) [inline cypress/support/cypress-commands/lifecycle.js:570]: Let's do a null check to be safe.

File: cypress/support/constants/constants.js [hunk 3]
```diff
@@ -400,7 +401,15 @@ module.exports = {
...
   APP_LAUNCH_STATUS: 'appLaunched',
+  VISIBILITYSTATE_VALIDATION_REQ: 'Lifecycle visibility state validation Req # ',
+  LIFECYCLE_VISIBILITYSTATE_SKIP_MESSAGE:
+    '`App is not reachable to fetch visibility state. Skipping Visibility state validation.`',
+  VISIBILITYSTATE_FAILURE_FIX_LOG:
+    '`, How to fix: If the visibility state response is correct for the current platform, but the expected value needs to be changed, add the expected value in configModule file ( cypress.config.js) in the format visibilityState: { ${state} : ${result.report}}`',
+  VISIBILITYSTATE_FAILURE_LOG:
+    '`. If the visibility state response is correct for the current platform, but the expected value needs to be changed, add the expected value in configModule file ( cypress.config.js) in the format visibilityState: { ${state} : ${result.report}}`',
   NO_EVENT_TRIGGERED: 'Expecting no event to be triggered from platform',
+  VISIBILITYSTATE: 'visibilityState',
```
Comment (kschrief) [inline cypress/support/constants/constants.js:412]: There are two "VISIBILITYSTATE" constants. Here and L356. Is this on purpose?

File: cypress/support/cypress-commands/lifecycle.js [hunk 4]
```diff
@@ -554,3 +555,70 @@ Cypress.Commands.add('lifecycleSchemaChecks', (response, state) => {
...
+    // Get validation requirements for the current scenario from the moduleReqId JSON
+    const scenarioRequirement = UTILS.getEnvVariable(CONSTANTS.SCENARIO_REQUIREMENTS);
+
+    // Fetching the requirement IDs for the "visiblilityState" from the scenarioRequirement.
+    const lifecycleStateRequirementId = scenarioRequirement.find((req) =>
+      req.hasOwnProperty('visible_check')
```
Comment (kschrief) [inline cypress/support/cypress-commands/lifecycle.js:576]: Let's add a "visible_check" constant if not already present

File: cypress/support/cypress-commands/lifecycle.js [hunk 5]
```diff
@@ -554,3 +555,70 @@ Cypress.Commands.add('lifecycleSchemaChecks', (response, state) => {
...
+    );
+
+    if (
```
Comment (kschrief) [inline cypress/support/cypress-commands/lifecycle.js:579]: The code in this if block is difficult to read/follow. Please put in line breaks to split up the logic as well as comments to explain what each set of lines does.

**Review by kschrief — CHANGES_REQUESTED** [review]

File: cypress/support/cypress-commands/lifecycle.js (hunk 5 above)
Comment (SnehalP251) [reply cypress/support/cypress-commands/lifecycle.js:581]: Added line breaks as well as comments.

File: cypress/support/cypress-commands/lifecycle.js (hunk 4 above)
Comment (SnehalP251) [reply cypress/support/cypress-commands/lifecycle.js]: Moved "visible_check" to constants.

File: cypress/support/constants/constants.js (hunk 3 above)
Comment (SnehalP251) [reply cypress/support/constants/constants.js:411]: Removed one "VISIBILITYSTATE" constants from L356

File: cypress/support/cypress-commands/lifecycle.js (hunk 2 above)
Comment (SnehalP251) [reply cypress/support/cypress-commands/lifecycle.js]: Added null check

File: cypress/support/cypress-commands/lifecycle.js (hunk 1 above)
Comment (SnehalP251) [reply cypress/support/cypress-commands/lifecycle.js]: Moved to "visibilityState" to contants

File: cypress/support/cypress-commands/lifecycle.js [hunk 6]
```diff
@@ -477,3 +478,67 @@ Cypress.Commands.add('lifecycleSchemaChecks', (response, state) => {
...
+        }
+        cy.log(
+          pretext + ': Expected : ' + visibilityState[state] + ' , Actual : ' + result.report
+        ).then(() => {
+          assert.equal(visibilityState[state], result.report, pretext);
+        });
+      });
+    }
+  } else {
+    cy.log(eval(CONSTANTS.LIFECYCLE_VISIBILITYSTATE_SKIP_MESSAGE));
```
[the remark this answers is not in the document]
Comment (SnehalP251) [reply cypress/support/cypress-commands/lifecycle.js]: Didn't find the better approach to resolve this.

File: cypress/support/cypress-commands/lifecycle.js [hunk 7]
```diff
@@ -477,3 +478,67 @@ Cypress.Commands.add('lifecycleSchemaChecks', (response, state) => {
...
+            assert(false, CONSTANTS.NO_MATCHED_RESPONSE);
+          });
+        } else if (result.error) {
+          assert(false, result.error.message);
+        }
+        const pretext =
+          CONSTANTS.VISIBILITYSTATE_VALIDATION_REQ + lifecycleStateRequirementId.visible_check.id;
+        if (visibilityState[state] != result.report) {
+          const fixLog = eval(CONSTANTS.VISIBILITYSTATE_FAILURE_FIX_LOG);
+          const assertLog = eval(CONSTANTS.VISIBILITYSTATE_FAILURE_LOG);
```
[the remark this answers is not in the document]
Comment (SnehalP251) [reply cypress/support/cypress-commands/lifecycle.js:613]: Didn't find the better approach to resolve this.

File: cypress/support/cypress-commands/lifecycle.js [hunk 8]
```diff
@@ -477,3 +478,67 @@ Cypress.Commands.add('lifecycleSchemaChecks', (response, state) => {
...
+          cy.log(CONSTANTS.NO_MATCHED_RESPONSE).then(() => {
+            assert(false, CONSTANTS.NO_MATCHED_RESPONSE);
+          });
+        } else if (result.error) {
+          assert(false, result.error.message);
+        }
+        const pretext =
+          CONSTANTS.VISIBILITYSTATE_VALIDATION_REQ + lifecycleStateRequirementId.visible_check.id;
+        if (visibilityState[state] != result.report) {
+          const fixLog = eval(CONSTANTS.VISIBILITYSTATE_FAILURE_FIX_LOG);
```
[the remark this answers is not in the document]
Comment (SnehalP251) [reply cypress/support/cypress-commands/lifecycle.js:612]: Didn't find the better approach to resolve this.

**Review by kschrief — APPROVED** [review]

**Pull request closed without merging**